### PR TITLE
Catch io.StdinException from failure to set stdin echo/line mode

### LINKED ISSUE #42284: Flutter Web (1.9+) + Docker -> Failure to launch
Hello Flutter community,

I've a question regarding Flutter Web (1.9+) version and Docker.

### Problem
I get an error that `No devices found with name or id matching 'chrome'`.

### Background
Previously I was running Flutter (1.7+) on Docker via `webdev` and it used to work like a charm (`webdev serve --release`). 
Now I updated my Flutter to latest version (Web was merged into flutter master). And started to face issues while trying to deploy it on my server (`flutter run -d chrome --release`).

### Setup
My `Dockerfile` content.

```
<...>
RUN curl -L https://storage.googleapis.com/flutter_infra/releases/stable/linux/flutter_linux_v1.9.1+hotfix.4-stable.tar.xz | tar -C /opt -xJ

RUN apt-get install -y lib32stdc++6

WORKDIR ${PROJECT_DIR}
COPY ./ ./

RUN flutter channel master
RUN flutter upgrade
RUN flutter config --enable-web
RUN flutter doctor
```

`flutter doctor` output:
```
Flutter 1.10.15-pre.16 • channel master • https://github.com/flutter/flutter.git
Framework • revision 14c1c211d5 (3 hours ago) • 2019-10-08 14:53:55 -0700
Engine • revision c635d70c72
Tools • Dart 2.6.0 (build 2.6.0-dev.6.0 f4a72bfc64)
Running flutter doctor...
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, v1.10.15-pre.16, on Linux, locale en_US)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/setup/#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, set ANDROID_HOME to that location.
      You may also want to add it to your PATH environment variable.
[!] Android Studio (not installed)
[!] Connected device
    ! No devices available
```
After running `flutter doctor` I can see my Flutter version is `1.10.15-pre.16` and I'm on `master`. However, I can also see 
```
[!] Connected device
    ! No devices available
```
Which is weird, because I should see Chrome as one of the devices?

Any help is appreciated!

### LINKED ISSUE #42384: OS Error: Inappropriate ioctl for device - Running Flutter from Docker
My pipeline requires to utilize docker to build apps utilizing different versions of flutter. It compiles and installs the APK successfully but after a couple of seconds after, it stops with the following error. The device is a Nexus 5 emulator connected by SSH to the docker container.

**Command**
```
sudo docker -i -v /projects/app1:/projects rubencg195/flutter-android-sdk bash -c "cd /projects && flutter run --pid-file ./pid.txt --observatory-port 8000 --disable-service-auth-codes lib/main.dart --enable-software-rendering"
```

_**StdinException: StdinException: Error setting terminal echo mode, OS Error: Inappropriate ioctl for device, errno = 25**_

The complete log is the following:

**Build Log**
```
Running Gradle task 'assembleDebug'...                          
flutter dependency
flutter dependency
flutter dependency
Running Gradle task 'assembleDebug'...                          
 31.5s
Built build/app/outputs/apk/debug/app-debug.apk.
Installing build/app/outputs/apk/app.apk...                     
12.7s
Syncing files to device Android SDK built for x86...            
D/EGL_emulation( 4460): eglMakeCurrent: 0xed88d800: ver 2 0 (tinfo 0xd022c810)
Syncing files to device Android SDK built for x86...            
7,953ms
 To hot reload changes while running, press "r". To hot restart (and rebuild state), press "R".
An Observatory debugger and profiler on Android SDK built for x86 is available at: http://127.0.0.1:8000/

For a more detailed help message, press "h". To detach, press "d"; to quit, press "q".
Sending crash report to Google.
Crash report sent (report ID: c3950bca910106ca)
Oops; flutter has exited unexpectedly.
Crash report written to /projects/flutter_02.log;
please let us know at https://github.com/flutter/flutter/issues.
```

**Output Error File**

```
Flutter crash report; please file at https://github.com/flutter/flutter/issues.

## command

flutter run --pid-file ./pid.txt --observatory-port 8000 --disable-service-auth-codes lib/main.dart --enable-software-rendering

## exception

StdinException: StdinException: Error setting terminal echo mode, OS Error: Inappropriate ioctl for device, errno = 25


#0      Stdin.echoMode= (dart:io-patch/stdio_patch.dart:84:7)
#1      AnsiTerminal.singleCharMode= (package:flutter_tools/src/base/terminal.dart:162:15)
#2      TerminalHandler.setupTerminal (package:flutter_tools/src/resident_runner.dart:992:14)
#3      RunCommand.runCommand.<anonymous closure> (package:flutter_tools/src/commands/run.dart:475:15)
#4      _rootRunUnary (dart:async/zone.dart:1132:38)
#5      _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#6      _FutureListener.handleValue (dart:async/future_impl.dart:137:18)
#7      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:678:45)
#8      Future._propagateToListeners (dart:async/future_impl.dart:707:32)
#9      Future._complete (dart:async/future_impl.dart:512:7)
#10     _SyncCompleter.complete (dart:async/future_impl.dart:51:12)
#11     HotRunner.attach (package:flutter_tools/src/run_hot.dart:208:26)
<asynchronous suspension>
#12     HotRunner.run (package:flutter_tools/src/run_hot.dart:267:12)
<asynchronous suspension>
#13     RunCommand.runCommand (package:flutter_tools/src/commands/run.dart:481:37)
<asynchronous suspension>
#14     FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:557:18)
#15     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:71:64)
#16     _rootRunUnary (dart:async/zone.dart:1132:38)
#17     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#18     _FutureListener.handleValue (dart:async/future_impl.dart:137:18)
#19     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:678:45)
#20     Future._propagateToListeners (dart:async/future_impl.dart:707:32)
#21     Future._completeWithValue (dart:async/future_impl.dart:522:5)
#22     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:30:15)
#23     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:288:13)
#24     RunCommand.usageValues (package:flutter_tools/src/commands/run.dart)
#25     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:71:64)
#26     _rootRunUnary (dart:async/zone.dart:1132:38)
#27     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#28     _FutureListener.handleValue (dart:async/future_impl.dart:137:18)
#29     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:678:45)
#30     Future._propagateToListeners (dart:async/future_impl.dart:707:32)
#31     Future._completeWithValue (dart:async/future_impl.dart:522:5)
#32     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:30:15)
#33     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:288:13)
#34     IosProject.isSwift (package:flutter_tools/src/project.dart)
#35     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:71:64)
#36     _rootRunUnary (dart:async/zone.dart:1132:38)
#37     _CustomZone.runUnary (dart:async/zone.dart:1029:19)
#38     _FutureListener.handleValue (dart:async/future_impl.dart:137:18)
#39     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:678:45)
#40     Future._propagateToListeners (dart:async/future_impl.dart:707:32)
#41     Future._completeWithValue (dart:async/future_impl.dart:522:5)
#42     Future._asyncComplete.<anonymous closure> (dart:async/future_impl.dart:552:7)
#43     _rootRun (dart:async/zone.dart:1124:13)
#44     _CustomZone.run (dart:async/zone.dart:1021:19)
#45     _CustomZone.runGuarded (dart:async/zone.dart:923:7)
#46     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:963:23)
#47     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#48     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#49     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:391:30)
#50     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:416:5)
#51     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
```

## flutter doctor

```
[✓] Flutter (Channel master, v1.10.15-pre.39, on Linux, locale en_US.UTF-8)
    • Flutter version 1.10.15-pre.39 at /opt/flutter
    • Framework revision 2adda80c10 (2 hours ago), 2019-10-09 16:58:41 -0700
    • Engine revision e693192167
    • Dart version 2.6.0 (build 2.6.0-dev.6.0 41d13f0ffe)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /opt/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /opt/sdk
    • Java binary at: /usr/local/openjdk-8/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_222-b10)
    • All Android licenses accepted.

[✗] Chrome - develop for the web (Cannot find chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/setup/#android-setup for detailed instructions).

[✓] Connected device (2 available)
    • Android SDK built for x86 • localhost:6000  • android-x86    • Android 8.1.0 (API 27) (emulator)
    • Headless Server           • headless-server • web-javascript • Flutter Tools

! Doctor found issues in 2 categories.
```

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Fixes https://github.com/flutter/flutter/issues/42384
Fixes https://github.com/flutter/flutter/issues/42284

Unclear why this isn't caught by the terminal check? I'll have to consult the runestones tomorrow.

Example crash:

```
Thread 0 main thread CRASHEDStdinException: StdinException: Error setting terminal echo mode, OS Error: Inappropriate ioctl for device, errno = 25

  | at Stdin.echoMode= | (stdio_patch.dart:84 )
-- | -- | --
  | at AnsiTerminal.singleCharMode= | (terminal.dart:162 )
  | at TerminalHandler.setupTerminal | (resident_runner.dart:992 )
  | at RunCommand.runCommand.<anonymous closure> | (run.dart:475 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._complete | (future_impl.dart:512 )
  | at _SyncCompleter.complete | (future_impl.dart:51 )
  | at ResidentWebRunner.run | (resident_web_runner.dart:202 )
  | at <asynchronous gap> | (async )
  | at RunCommand.runCommand | (run.dart:481 )
  | at <asynchronous gap> | (async )
  | at FlutterCommand.verifyThenRunCommand | (flutter_command.dart:557 )
  | at _asyncThenWrapperHelper.<anonymous closure> | (async_patch.dart:71 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._completeWithValue | (future_impl.dart:522 )
  | at _AsyncAwaitCompleter.complete | (async_patch.dart:30 )
  | at _completeOnAsyncReturn | (async_patch.dart:288 )
  | at RunCommand.usageValues | (run.dart )
  | at _asyncThenWrapperHelper.<anonymous closure> | (async_patch.dart:71 )
  | at _rootRunUnary | (zone.dart:1132 )
  | at _CustomZone.runUnary | (zone.dart:1029 )
  | at _FutureListener.handleValue | (future_impl.dart:137 )
  | at Future._propagateToListeners.handleValueCallback | (future_impl.dart:678 )
  | at Future._propagateToListeners | (future_impl.dart:707 )
  | at Future._completeWithValue | (future_impl.dart:522 )
  | at _AsyncAwaitCompleter.complete | (async_patch.dart:30 )
  | at _completeOnAsyncReturn | (async_patch.dart:288 )


```

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Test=???

File: packages/flutter_tools/lib/src/base/io.dart [hunk 1]
```diff
@@ -181,6 +181,36 @@ class Stdio {
...
+  ///
+  /// [io.Stdin.hasTerminal] only covers a subset of cases. In this check the
+  /// echoMode is toggled on and off to catch cases where the tool running in
+  /// a docker container thinks there is an attached terminal. This can cause
+  /// runtime errors such as "inappropriate ioctl for device" if not handled.
+  bool get stdinHasTerminal {
+    if (_stdinHasTerminal != null) {
+      return _stdinHasTerminal;
+    }
+    if (stdin is io.Stdin) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/io.dart]:
```dart
if (stdin is! io.Stdin) {
  return _stdinHasTerminal = false;
}
```
then unindent.

File: packages/flutter_tools/lib/src/base/terminal.dart [hunk 2]
```diff
@@ -155,8 +155,8 @@ class AnsiTerminal {
   String clearScreen() => supportsColor ? clear : '\n\n';
 
   set singleCharMode(bool value) {
-    final Stream<List<int>> stdin = io.stdin;
-    if (stdin is io.Stdin && stdin.hasTerminal) {
+    if (io.stdio.hasTerminal) {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/terminal.dart]: `io.stdinHasTerminal` ?
Also, you could flip the sense of the test and return early then unindent.

File: packages/flutter_tools/lib/src/base/io.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/io.dart]: Done

File: packages/flutter_tools/lib/src/base/terminal.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/terminal.dart]: Done

File: packages/flutter_tools/lib/src/base/io.dart [hunk 3]
```diff
@@ -181,6 +181,36 @@ class Stdio {
...
+    final io.Stdin ioStdin = stdin;
+    if (!ioStdin.hasTerminal) {
+      return _stdinHasTerminal = false;
+    }
+    try {
+      final bool currentEchoMode = ioStdin.echoMode;
+      ioStdin.echoMode = !currentEchoMode;
+      ioStdin.echoMode = currentEchoMode;
+    } on io.StdinException {
+        return _stdinHasTerminal = false;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/io.dart]: this looks indented too much.

**Review by zanderso — APPROVED** [review]

File: packages/flutter_tools/lib/src/base/io.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/io.dart]: Fixed

**Comment by jonahwilliams** [comment]
Closing in favor of https://github.com/flutter/flutter/pull/43225 due to CI weirdness

**Pull request closed without merging**